File: Divide and Conquer Algorithm/Maximum subarray(contiguous) problem/Kadane's Algorithm (Optimal Solution)/kadane_algorithm.py
# function for finding maximum sum of contiguous array in an Array

def maxArraySum(arr):
    global_max = curr_max = arr[0]

    for i in range(1, len(arr)):
        
        curr_max = max(curr_max, curr_max + arr[i])
        
        if(curr_max > global_max):
            global_max = curr_max

    return global_max



arr = [-2, 3, 2, -1]
maxSum = maxArraySum(arr)
print('The maximum contiguous array sum is :',maxSum)

File: Divide and Conquer Algorithm/Maximum subarray(contiguous) problem/Kadane's Algorithm (Optimal Solution)/kadane_algorithm.c
#include<stdio.h>

int maxSubArraySum(int arr[], int arr_size)
{
    int global_max = arr[0], curr_max = arr[0];
    for (int i = 1; i < arr_size; i++)
    {
        curr_max = curr_max + arr[i];
        if(curr_max < arr[i]) {
            curr_max = arr[i] ;
        }

        if (global_max < curr_max) {
            global_max = curr_max ;
        }
    }
    return global_max ;
}


int main()
{
    int arr[] = {-2, -3, 4, -1, -2, 1, 5, -3};
    int arr_size = sizeof(arr)/sizeof(arr[0]); // way of determining the size of array
    int max_sum = maxSubArraySum(arr, arr_size);
    printf("Maximum contiguous sum is : %d", max_sum);
    return 0;
}
